src/cpu/flexcpu/simple_dataflow_thread.cc

```diff
@@ -53,19 +53,63 @@ SDCPUThread::SDCPUThread(SimpleDataflowCPU* cpu_, ThreadID tid_,
     //
 }
 
-ThreadContext* SDCPUThreadState::getThreadContext()
+ThreadContext*
+SDCPUThread::getThreadContext()
 {
     return _committedState.getTC();
 }
 
 ExecContext*
-SDCPUThreadState::getExecutionContext()
+SDCPUThread::getExecutionContext()
 {
     return &_isaInterface;
 }
 
 void
-SDCPUThreadState::startup()
+SDCPUThread::onBeginFetch()
+{
+    // Retrieve the value of the speculative PC as a raw Addr value.
+    Addr inst_vaddr = _speculativePC.instAddr();
+
+    RequestPtr fetch_req = std::make_shared<Request>();
+
+    fetch_req->taskId(_cpuRef->taskId()); // Necessary? TimingSimpleCPU seems
+                                          // to do this...
+    // TODO continue setting up the request before sending it
+
+    if (!_cpuRef->requestInstAddrTranslation(fetch_req,
+        [this](Fault f, RequestPtr& r) {onPCTranslated(f, r);})) {
```

Comment:
I feel less excited about the function pointers now. Will the CPU ever call a different function other than `onPCTranslated` after completing this request? If not, I think it would be clearer to pass in `this` and have the CPU call `this->onPCTranslated`. By using a function pointer here, you're hiding the context of the thread.

However, I'm willing to wait this out and see where it goes. I think it could be interesting for the CPU to be completely agnostic of the thread type, but I'm not sure if this characteristic will be used. If it's not used, then I'm not sure it's worth it.

One final thought, I do like how it, in some sense, keeps the flow control information local. It's obvious from reading this thread code that after you call `requestInstAddrTranslation` next `onPCTranslated` will be called.